test/plugin/test_out_google_cloud_grpc.rb

```diff
@@ -0,0 +1,310 @@
+  class GRPCLoggingMockFailingService <
+      Google::Logging::V1::LoggingService::Service
+    # 'code_sent' and 'message_sent' are references of external variables. We
+    #  will assert the values of them later. 'code_value' and 'message_value'
+    #  are actual error code and message we expect this mock to return.
+    def initialize(code, message, failed_attempts)
+      @code = code
+      @message = message
+      @failed_attempts = failed_attempts
+      super()
+    end
+
+    def write_log_entries(_request, _call)
+      @failed_attempts << 1
+      fail GRPC::BadStatus.new(@code, @message)
+    end
+
+    # TODO(lingshi) Remove these dummy methods when grpc/9033 is fixed.
+    #
+    # These methods should never be called, so they will just fail the tests
+    # with "unimplemented" errors..
+    def _undefined
+      fail "Method #{__callee__} is unimplemented and needs to be overridden."
+    end
+
+    alias_method :list_logs, :_undefined
+    alias_method :list_log_services, :_undefined
+    alias_method :list_log_service_indexes, :_undefined
+    alias_method :delete_log, :_undefined
+    undef_method :_undefined
+  end
+
+  # Set up grpc stubs to mock the external calls.
+  def setup_logging_stubs(should_fail = false, code = 0, message = 'Ok')
+    srv = GRPC::RpcServer.new
+    @failed_attempts = []
+    @requests_sent = []
+    if should_fail
+      grpc = GRPCLoggingMockFailingService.new(code, message, @failed_attempts)
+    else
+      grpc = GRPCLoggingMockService.new(@requests_sent)
+    end
+    srv.handle(grpc)
+    srv.add_http2_port(GRPC_MOCK_HOST, :this_port_is_insecure)
+    t = Thread.new { srv.run }
+    srv.wait_till_running
+    begin
+      yield
+    rescue Test::Unit::Failure, StandardError => e
+      srv.stop
+      t.join
+      raise e
+    end
+    srv.stop
+    t.join
+  end
+
+  # Verify the number and the concent of the log entries match the expectation.
+  # The caller can optionally provide a block which is called for each entry.
+  def verify_log_entries(n, params, payload_type = 'textPayload', &block)
+    @requests_sent.each do |batch|
+      @logs_sent << JSON.parse(batch.to_json)
+    end
+    verify_json_log_entries(n, params, payload_type, &block)
+  end
+
+  # For an optional field with default values, Protobuf omits the field when it
+  # is deserialized to json. So we need to add an extra check for gRPC which
+  # uses Protobuf.
+  #
+  # An optional block can be passed in if we need to assert something other than
+  # a plain equal. e.g. assert_in_delta.
+  def assert_equal_with_default(field, expected_value, default_value, entry)
+    if expected_value == default_value
+      assert_nil field
+    elsif block_given?
+      yield
+    else
+      assert_equal expected_value, field, entry
+    end
+  end
+
+  # A wrapper around the constant HTTP_REQUEST_MESSAGE, so the definition can be
+  # skipped in the shared module here and defined in the test class later.
```

Comment:
This comment doesn't make sense here -- you're already *in* the test class... How about:
```
# A wrapper around the constant HTTP_REQUEST_MESSAGE, so the definition can be
# skipped in the shared module and defined here.
```
?